spi-proposal.md

```diff
@@ -0,0 +1,140 @@
+SPI
+===
+
+A proposal for a SPI protocol for Firmata.
+
+Each SPI device is initialized via a config message. This message sets the bit order
+data mode clock speed and optional CS pin for the device. Arrays of bytes can be
+transferred to and received from the slave device. By default the CS pin pulled low
+at the beginning of a transfer and released after the last byte sent. The user can
+change the active CS pin state or even opt to handle the CS pin manually.
+
+Additional options can be set for less common cases such as changing the word length
+or forcing the CS pin to toggle between every word sent.
+
+There are 3 ways to send and receive data from the SPI slave device:
+
+1. Transfer - for each word sent a word is received
+2. Write - only write data (ignore any data retuned by the slave device)
+3. Read - only read data, writing 0 for each byte to be read
+
+
+## Protocol details
+
+### Begin
+
+```
+// called once to initialize SPI hardware
+// can optionally be set internally on first SPI_CONFIG_DEVICE message
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_BEGIN (bits 0 - 4) | channel (bits 5 - 6) if HW supports multiple SPI ports
+3:  END_SYSEX
+```
+
+### Config
+
+Data modes:
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+```
+// sent once for each SPI device
+0:  START_SYSEX
+1:  SPI_DATA
+2:  SPI_CONFIG_DEVICE (bits 0 - 4) | channel (bits 5 - 6) if HW supports multiple SPI ports
```

Comment:
Split channel out into a separate byte, always reserve the full range of 127 bits for the command ID (SPI_CONFIG_DEVICE in this case) to ensure scalability.